Native media Equirect layers creation

danrossi opened the pull request:
Related issue: #31029

**Description**

Adding native media layer creation for WebGPU XR. This fixes a context loss issue doing it outside the manager. There is a potential Oculus browser bug complaining makeXRCompatible isn't set when setting up a media layer still. 

Media layers are prepended into the background before the other layers. 

Example https://github.com/danrossi/three.js/blob/equirect-layers/examples/webgpu_xr_media_layer.html

danrossi committed:
Adding native media layer creation for WebGPU XR. This fixes a context loss issue doing it outside the manager. #31029

danrossi committed:
cleanup example

danrossi commented:
It took a bit to design how to go about it. The only issue is you have to toggle the video mesh layers when going into XR and ending the session. Maybe there is other ways to do it.

https://github.com/danrossi/three.js/blob/equirect-layers/examples/webgpu_xr_media_layer.html#L130
https://github.com/danrossi/three.js/blob/equirect-layers/examples/webgpu_xr_media_layer.html#L151



live example https://electroteque.org/dev/threejs/examples/webgpu_xr_media_layer.html

danrossi committed:
fix linting

danrossi committed:
add a layers check

Mugen87 committed:
Update XRManager.js

Improve docs.

Mugen87 commented:
I have tried to improve the JSDoc a bit. For a functional review I hope @cabanier can have a look.

Mugen87 commented on examples/webgpu_xr_media_layer.html:
```diff
@@ -0,0 +1,185 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js vr - 360 stereo video</title>
```
Please include the new example in [files.json](https://github.com/mrdoob/three.js/blob/dev/examples/files.json). Otherwise it won't appear on the homepage.

danrossi reviewed (commented):
looks ok

danrossi commented:
The requirement to toggle off the video texture with the meshes. I'm not sure how to make better. It causes a crash if the texture is also rendered.

danrossi committed:
add webgpu_xr_media_layer example with screenshot

danrossi committed:
fix examples data file

cabanier commented on src/renderers/common/XRManager.js:
```diff
@@ -596,6 +612,40 @@ class XRManager extends EventDispatcher {
+	 * @param {boolean} [is180=false] If it's a 180 video.
+	 * @param {Object} [params={}] Extra params for the layer to add but not needed.
+	 */
+	createMediaLayer( video, layout = 'mono', transform = {}, is180 = false, params = {} ) {
```
I don't think this code implements emulation for 2D or browser that don't support xr layers. It will also not dynamically add them while in VR.
Please look at the flow of the quad and cylinder layer.

danrossi replied:
It does as normal with the meshes added to the scene, just when going to XR the mesh layers need to be disabled, if it doesn't there is a crash in the Oculus browser. I can't get my head around those methods and there is no texture render required. 

There is specific logic for the stereo and mono video that shouldn't be in the manager perhaps. There is multiple meshes for the stereo version. 

So even if it was created in the manager and added to the scene  the meshes need to be disabled. 

The native layer only gets setup if layers are supported. Works with the Rift in XR that doesn't support layers it gets rendered in the projection layer. 

https://github.com/danrossi/three.js/blob/equirect-layers/examples/webgpu_xr_media_layer.html#L129

https://electroteque.org/dev/threejs/examples/webgpu_xr_media_layer.html

I could automatically disable/enable the mesh layers perhaps. That part I find is little clunky.  There is nothing for it to do in the `renderLayers` method unless it's non layers and not in XR.

If you have another idea it would be good.

cabanier replied:
The code that creates the video layer in xr should also the emulation in 2D mode. Right now you create a video texture in your sample code and disable it on session start. That logic needs to go to xrmanager.
This code also doesn't work if you add a media layer while in VR.

danrossi replied:
So create the single or multiple mesh in the method. Return and add the geometries before adding to the scene. the issue is while in XR. Is a special render target needed like the others where its then only rendering when not in XR ? so a different render method is needed to the `renderLayers` method. To prevent that layer disable hack needed. I'm assuming that is the requirement if it actually works like that.

danrossi replied:
I've done further testing in my implementation and to switch between projection layouts while in XR for video playlists. It needs tricks to replace and swap out the layer at a specific index. The properties on the layer are read only so can't be switched like that so a new layer needs to be created. There was a bug where projection changes was prepending new layers causing eventual crashes. Swapping layers too quickly and frequent may still cause a crash.

cabanier reviewed (commented):
Please match behavior of quad/cylinder layers

danrossi committed:
Add 2D media layer creation into XR api and return a group
Move equirect layer creation into common layer creation method
Fixup example with new media layer usage

danrossi committed:
update media layer example screenshot

danrossi commented:
Is this looking good setting the geometry externally. 
 
 I have it working implementing and tested. Unfortunately media layer needs to be added in the manager as doing it external caused a context loss fault. It may be a problem adding other layers externally as the media layers list needs to be prepended in the list as backgrounds as I have done. 
 
 The only issue still I can't figure out when going from the 2D mesh to native layer in XR. It will be rendered including the video texture in the projection layer as its added to the scene so those layers need to be turned off. Or maybe simply removing those mesh groups from the scene ?. I can look at doing it in the manager perhaps. Turn off mesh layers for all the media layers in the list on session start. Unless there is a better way to not render those meshes in the scene into the projection layer.

cabanier commented:
> ... doing it external caused a context loss fault.

Do you have an example for me to look at? You shouldn't be getting context losses.

>  I can look at doing it in the manager perhaps. Turn off mesh layers for all the media layers in the list on session start. Unless there is a better way to not render those meshes in the scene into the projection layer.

Yes, can't you do the same as what I did for quad/cylinder layer?

danrossi reviewed (commented):
looks ok

danrossi commented:
If the external geometry setter is ok. Or I make them default SphereGeometry requiring to update it's settings. I just set them to PlaneGeometry.

https://github.com/danrossi/three.js/blob/equirect-layers/examples/webgpu_xr_media_layer.html#L105

I add the mesh group to the layers params. 

https://github.com/danrossi/three.js/blob/equirect-layers/src/renderers/common/XRManager.js#L667

So this requirement to turn off texture rendering in the projection layer can be done internally perhaps and make it transparent. 

https://github.com/danrossi/three.js/blob/equirect-layers/examples/webgpu_xr_media_layer.html#L114

danrossi commented:
It might need a blank render target like the other layer examples. When using immersive-vr mode the projection layer render target blocks out the background native media layer. So it seems it's a blank render.

https://github.com/mrdoob/three.js/blob/dev/src/renderers/common/XRManager.js#L1480
https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl-fallback/WebGLBackend.js#L354

https://electroteque.org/dev/threejs/examples/webgpu_vr_media_layer.html

danrossi commented on examples/webgpu_xr_media_layer.html:
```diff
@@ -0,0 +1,167 @@
+
+				scene.add(mediaLayer);
+
+				renderer.xr.addEventListener("sessionstart", async () => {
```
Do you mean adding the group to the scene or managing the 2D layers internally ? Adding to the scene is external like the other demo.  I may need to look at a render option for immersive-vr as it's a black render.

danrossi replied:
If I understood correctly place the 2D mesh toggling internally. As the group adding to the scene is external like the other layer example. Example updated. 

There is still an issue with immersive-vr mode and may need that render target code in the frame loop. 

I added the XR controllers to show the projection layer render however the model is displaying artifacts for some reason. It's ok on the dev branch.  

https://electroteque.org/dev/threejs/examples/webgpu_xr_media_layer.html

danrossi replied:
I just worked out this isn't possible with an equirect layer. So in immersive-vr mode the native layer is being hidden by the three render. So not sure what method to use to "punch through". It's ok for immersive-ar mode. 

```
 const glSubImage = this._glBinding.getSubImage( layer2.xrlayer, this._xrFrame );
 ````

danrossi committed:
Manage 2D media mesh layer toggling within the XRManager
Tidy up media layer exanmple and add XR Controllers displayed on the projection layer

danrossi committed:
Merge branch 'dev' into equirect-layers

danrossi committed:
fix merged conflict issue

danrossi commented:
I was able to figure out fixing the conflict in the editor but have been testing on a merged branch. I'm still unable to figure out how to work around three projection layer render target hiding the native layer in immersive-vr mode. Only works in immersive-ar mode. Which is what all the XR examples use. 

Hopefully disabling the 2D mesh layer internally to not render in the projection layer is the right solution.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

cabanier commented on examples/webgpu_xr_media_layer.html:
```diff
@@ -0,0 +1,223 @@
+
+				// right
+
+				const geometry2 = new THREE.SphereGeometry( 500, 60, 40 );
```
it seems this should go into the XRManager like we do for the other layer types.

danrossi replied:
Ok I will add extra params for that. Not a problem. I'm stuck personally how to work around the projection layer render target masking out the background native layer when in immersive-vr mode. With your other layer examples, your method allows for a dummy render target from the looks of it. To allow to see the native layer but not with the equi layer option.

cabanier commented on src/renderers/common/XRManager.js:
```diff
@@ -601,7 +618,88 @@ class XRManager extends EventDispatcher {
+	 * @param {Object} [params={}] Extra params for the layer to add but not needed.
+	 * @returns {Group} Returns a group of a mono or stereo mesh
+	 */
+	createMediaLayer( texture, layout = 'mono', quaternion = {}, is180 = false, params = {} ) {
```
should you add an option to make it stereo?

danrossi committed:
Move SphereGeoemtry into the createMediaLayer method.
Add uv factor mappings for both stereo-left-right and stereo-top-bottom layouts for each eye index. For the geometry uv transformations.
Update media layer example with some extra comments on usage.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Add option for stereo layout input which defaults to stereo-top-bottom

danrossi commented on examples/webgpu_xr_media_layer.html:
```diff
@@ -0,0 +1,223 @@
+				renderer.xr.enabled = true;
+				renderer.xr.setReferenceSpaceType( 'local' );
+
+				const mediaLayer = renderer.xr.createMediaLayer(texture, 'stereo-left-right', { x: 0, y: .28, z: 0, w: .96 });
```
It's just setting up an equirectLayer. Non layer WebXR goes to the projection layer from the mesh group added to the scene. So handled by three. Which is disabled from being rendered with the native layer ?

https://github.com/danrossi/three.js/blob/equirect-layers/src/renderers/common/XRManager.js#L1674

There is no texture for the native layer and a mediabinding layer, you can't do this. It caused an error when I tried.

https://github.com/danrossi/three.js/blob/equirect-layers/src/renderers/common/XRManager.js#L967

danrossi replied:
I've just run into an issue with the default immersive-ar mode used in the examples and required by the media layer feature for now. When switching videos in XR it flashes to the camera during load and metadata stage. There is other occurrences with 180 video where the camera in the areas not covered up by the native layer is being displayed and needs a skybox behind it. 

Until something is figured out to render a blank stream target to get media layers appearing in immersive-vr mode. A blank skybox needs to appear during load for the projection layer background so it covers up the camera. Then hide when beginning playback. Setting a scene background color doesn't help. It needs to be a sphere mesh with a color set.

I'll have to figure out the immersive-vr mode support as the ar mode is annoying and not useful. 

Pseudo code for when in ar immersive mode.

```
if (!this.questSkybox) {
                    this.questSkybox = WebVRUtils.createBlankSkybox();
                    this.scene.children.splice(0, 0, this.questSkybox);
    
                    this.questSkybox.visible = false;
                 
                }
    
                this.questSkybox.visible = true;
                
                const onFirstFrame = () => {
                    this.questSkybox.visible = false;
                    this.video.removeEventListener("play", onFirstFrame);
                };

                this.video.addEventListener("play", onFirstFrame);
    ```

danrossi replied:
I found a solution for the rendering issue in immersive-vr but it may be crude and need extra work. It partially worked for 180 video only one side was appearing. If I added the same geometry into an empty mesh with no texture to the group and render that in XR on the scene while turning off the other meshes. It "poked through" to the native layer. 

It doesn't require creating a quad layer and then extra calls to "renderLayers" although I tried that. On exit those empty mesh layers can be turned off.  Will keep testing.

danrossi replied:
I figured out a solution for immersive-vr mode and keeping the example vr mode as ar is not useful and displays wrong. It's similar to the other layers render but it doesn't require extra external calls which didn't work anyway.

Creating blend meshes using the same geometry and enable those while in XR shows through to the native layer while doesn't affect anything rendered on the projection layer. I hope this is the right way and performs ok. It's blank textureless blend sphere meshes. Works for stereo, stereo 180 and mono.

https://github.com/danrossi/three.js/blob/equirect-layers/src/renderers/common/XRManager.js#L736

https://electroteque.org/dev/threejs/examples/webgpu_xr_media_layer.html

danrossi committed:
update code doc

danrossi committed:
make the default media layer layout 'stereo; which defaults to a 'stereo-top-bottom' layout.
update code docs to suit.

danrossi committed:
update native media layer example to support stereo and mono video demos.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Add 180 demo source

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Make media layer example use immersive-vr sessions as AR sessions show the camera which displays wrong.
Fixup media layer feature to render blend spheres while in XR to poke through the projection layer to properly show the native media layers while in immersive-vr mode.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
fix issue with replacing the current native media layer with a different layout or 180 content while in a xr session. Prevent prepenmding new layers and replace layers at specific indexes.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi commented:
I'm still maintaining this and syncing fork. However the project I did this for is on hold due to the native layers crashing the browser very badly and not stable. Its a potential memory leak and causes 100% cpu spikes then crashes, reporting the crashes has done thing. 

It seems unfit for purpose and not ready for production sadly. It performs better than with multiview in the projection layer however. But changing videos and swapping native layers to change the layout or angle between 360 and 180, and even exiting XR will make it crash badly eventually. Its cost me 30+ hours I cant recover making tests and reporting. And it's still a problem for a few months now. 

I may have to convince them to use the non native layers configuration for as long as it takes after spending time integrating it all. To keep moving forward. But the project is also on hold also because of the regression 5.7k h265 unable to be played in the browser when i spent a month beforehand testing it then broke after updates, disrupting things. I can't test any of their content at all now in Quest 2 v78

danrossi committed:
Add quest performance hit fix. Remove video element from DOM. Remove canvas container from DOM while in XR and return back to DOM on XR exit.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi commented:
I fixed up the demo to improve Quest performance hit issues. Remove video element from DOM and remove the whole container and canvas from the DOM while in XR and return after. 

The large video still causes the OS to be sticky with movements and fan ramps up when exiting XR. So the layers are not clearing when exiting XR and causing cpu spikes it seems. Definitely a memory leak is happening I watch the memory drop when re-entering and exiting XR. 

https://electroteque.org/dev/threejs/examples/webgpu_xr_media_layer.html

cabanier commented:
> I fixed up the demo to improve Quest performance hit issues. Remove video element from DOM and remove the whole container and canvas from the DOM while in XR and return after.
> 
> The large video still causes the OS to be sticky with movements and fan ramps up when exiting XR. So the layers are not clearing when exiting XR and causing cpu spikes it seems. Definitely a memory leak is happening I watch the memory drop when re-entering and exiting XR.
> 
> https://electroteque.org/dev/threejs/examples/webgpu_xr_media_layer.html

Quest browser 39.3 will release soon with a potential fix for this.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi commented:
The native media layer is finally stable as of V40.0 with a recent browser update. 

this shouldn't crash from a memory leak going in and out of XR. 

https://electroteque.org/dev/threejs/examples/webgpu_xr_media_layer.html

danrossi committed:
Merge branch 'dev' into equirect-layers

danrossi committed:
Merge branch 'equirect-layers' of github.com:danrossi/three.js into equirect-layers

danrossi commented:
I just updated to v79 and tested the merged changes and I now get flashing over the layer. Which is possibly the blend layer. I don't see the problem when integrated so it's just with the example.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'equirect-layers' of github.com:danrossi/three.js into equirect-layers

danrossi committed:
fix media layer creation with refactored layers support flag
cleanup media layer example

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
correct the default layer eye index to support mono videos

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
clean up example. Use VRButton to prevent monochrome camera showing up in Quest 2.

danrossi commented:
Everything is stabilised and operating correctly. This feature is needed in Quest if using multiview due to a bug #32151 I reported using multiview with  video textures. It causes flickering. So for Quest native media layer can be used instead.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Adding media quad native layer feature
Move common mesh and material utilities to a seperate module
Reuse the mesh material for the second eye for stereo
Update example with quad video samples.

danrossi committed:
cleanup example

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
fix blend layer position

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
/native quad layer units is double required than world units. Blend later transform is in set world units.

danrossi committed:
diovide the units of the quad native layer by half to match world units. Z transform can stay the same.

danrossi committed:
cleanup uneccessary blend layer group

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi commented:
I added the media layer feature almost 12 months ago. Is it possible to check if the design is correct and if useful. Its stable for me now. But for WebGPU support XRManager requires a factory method to handle layers like the WebGPURenderer system.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi committed:
Add webgpu module export of mesh and material helper methods used for layers creation.

danrossi committed:
Merge branch 'mrdoob:dev' into equirect-layers

danrossi commented:
XRManager changes won't merge in github where it lets me edit it. Sourcetree is always buggy for these things and won't stage chunks after editing the conflict. Not an expert to recover it properly. I have to start a fresh branch and PR so will close this. I made common helper methods for all the layer creation I need to reapply. Prettier royally messes with the lint so is disabled for this project.
